Newsletter categories: exposes value of wpcom_newsletter_categories_location to JavaScript

TimBroddin opened the pull request:
Fixes https://github.com/Automattic/wp-calypso/issues/82322

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion
<!-- If you're an Automattician, include a shortlink to the p2 discussion with Jetpack Product here. -->
<!-- Make sure any changes to existing products have been discussed and agreed upon -->

## Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->

## Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

- Follow the testing instructions below to apply to your sandbox
- Create a new post
- Inspect the `post-new.php` iframe in your devtools:

![CleanShot 2023-09-28 at 11 30 02@2x](https://github.com/Automattic/jetpack/assets/528287/1b1a0547-db14-4694-a4aa-c0b2d1922479)

- Output the value of `Jetpack_Subscriptions` it should be `block`
- Insert these lines inside `0-sandbox.php` and refresh:

```
function wpcom_newsletter_categories_location() {
	return 'modal';
}

add_filter( 'wpcom_newsletter_categories_location', 'wpcom_newsletter_categories_location', 10 );
```

- `modal` should now be returned

TimBroddin committed:
This update exposes the Newsletter category settings load to JavaScript. The settings allow the system to determine the location of the newsletter categories, which can now be modified using the 'wpcom_newsletter_categories_location' filter.

mashikag commented on projects/packages/jetpack-mu-wpcom/src/class-jetpack-mu-wpcom.php:
```diff
@@ -136,6 +138,26 @@ public static function load_map_block_settings() {
 	}
 
+	/**
+	 * Adds a global variable containing the map provider in a map_block_settings object to the window object.
```
I believe this comment is a copy/paste left over. 😁

TimBroddin replied:
Shoot! Fixed it.

TimBroddin committed:
Update comments for load_newsletter_categories_settings function

Changed the purpose of the function load_newsletter_categories_settings in the comments. It now indicates that the function is to store the information about where the newsletter categories should be shown, not the map provider information as previously stated. The code change reflects this new focus.

mashikag commented on projects/packages/jetpack-mu-wpcom/src/class-jetpack-mu-wpcom.php:
```diff
@@ -136,6 +138,26 @@ public static function load_map_block_settings() {
 	}
 
+	/**
+	 * Adds a global variable containing the where the newsletter categories should be shown.
```
I am sorry for being pesky about the comment, but I believe you missed a word: "containing the where the newsletter" 🙈

TimBroddin replied:
I'm moving too fast :D Fixed it.
Thanks for your patience 🙏

TimBroddin committed:
Update comments for load_newsletter_categories_settings function

Changed the purpose of the function load_newsletter_categories_settings in the comments. It now indicates that the function is to store the information about where the newsletter categories should be shown, not the map provider information as previously stated. The code change reflects this new focus.

mashikag commented:
BTW, I looked up the documentation on the `wp_localize_script`, and this is what it said ([here](https://developer.wordpress.org/reference/functions/wp_localize_script/#:~:text=Though%20localization%20is,to%20localize%20strings.)):

> Though localization is the primary use, it was often used to pass generic data from PHP to JavaScript, because it was originally the only official way to do that. [wp_add_inline_script()](https://developer.wordpress.org/reference/functions/wp_add_inline_script/) was [introduced in WordPress Version 4.5](https://make.wordpress.org/core/2016/03/08/enhanced-script-loader-in-wordpress-4-5/), and is now the best practice for that use case. `[wp_localize_script()](https://developer.wordpress.org/reference/functions/wp_localize_script/) ` should only be used when you actually want to localize strings.

I see the `wp_add_inline_script` is already used in the same file ([here](https://github.com/Automattic/jetpack/blob/39b239bdf7b1d5c059f0320a90fc35f8d6a6fbef/projects/packages/jetpack-mu-wpcom/src/class-jetpack-mu-wpcom.php#L164)). Is there a particular reason why you chose to use `wp_localize_script` instead? 🤔

mashikag reviewed (approved):
Either way I do not believe my above remark should be PR blocking. Other than that the code looks good, and works as expected. 👍

TimBroddin commented:
> BTW, I looked up the documentation on the `wp_localize_script`, and this is what it said ([here](https://developer.wordpress.org/reference/functions/wp_localize_script/#:~:text=Though%20localization%20is,to%20localize%20strings.)):
> 
> > Though localization is the primary use, it was often used to pass generic data from PHP to JavaScript, because it was originally the only official way to do that. [wp_add_inline_script()](https://developer.wordpress.org/reference/functions/wp_add_inline_script/) was [introduced in WordPress Version 4.5](https://make.wordpress.org/core/2016/03/08/enhanced-script-loader-in-wordpress-4-5/), and is now the best practice for that use case. `[wp_localize_script()](https://developer.wordpress.org/reference/functions/wp_localize_script/) ` should only be used when you actually want to localize strings.
> 
> I see the `wp_add_inline_script` is already used in the same file ([here](https://github.com/Automattic/jetpack/blob/39b239bdf7b1d5c059f0320a90fc35f8d6a6fbef/projects/packages/jetpack-mu-wpcom/src/class-jetpack-mu-wpcom.php#L164)). Is there a particular reason why you chose to use `wp_localize_script` instead? 🤔

`wp_localize_script` was recommended to me by @jeherve I think while working on the Maps block. It seems to be a clean way to expose values from PHP to JS, although I understand that we're probably not using it for what it was meant for 🙊

mashikag commented:
BTW, I run into an issue when testing, or at least I thought there was an issue. In Calypso I had "Classic View" selected for the posts page. As it comes out, this results in the post editor being loaded in a different manner (even when using the same **Add new post** submenu item). As a result there was no `new-post.php` iFrame present (or at least it is not referred to as such in the dev tools) but instead the values are found in the `top` iFrame. 👍

TimBroddin commented:
> BTW, I run into an issue when testing, or at least I thought there was an issue. In Calypso I had "Classic View" selected for the posts page. As it comes out, this results in the post editor being loaded in a different manner (even when using the same **Add new post** submenu item). As a result there was no `new-post.php` iFrame present (or at least it is not referred to as such in the dev tools) but instead the values are found in the `top` iFrame. 👍

Nice catch! Glad to hear that it also works correctly without an iframe.
